amplify/backend/function/getMembersForColony/src/index.js

```diff
@@ -0,0 +1,202 @@
+    // any reputation, and we don't want to break the whole query
+  }
+
+  /*
+   * Validate Colony addresses
+   */
+  let checksummedAddress;
+  try {
+    checksummedAddress = utils.getAddress(colonyAddress);
+  } catch (error) {
+    throw new Error(
+      `Colony address "${colonyAddress}" is not valid (after checksum)`,
+    );
+  }
+
+  // get list of watchers for colony
+  let watchers = [];
+  if (domainId === ALL_DOMAIN_ID || domainId === ROOT_DOMAIN_ID) {
+    const { data, errors } = await graphqlRequest(
+      getWatchersInColony,
+      { id: checksummedAddress },
+      GRAPHQL_URI,
+      API_KEY,
+    );
+
+    if (errors || !data) {
+      const [error] = errors;
+      throw new Error(
+        error?.message || 'Could not fetch colony data from DynamoDB',
+      );
+    }
+
+    // Remove members that have reputation (i.e. contributors) to
+    // leave only watchers
+    watchers = data?.getColonyByAddress?.items[0]?.watchers.items.filter(
+      (item) =>
+        addressesWithReputation?.every(
+          (address) => address.toLowerCase() !== item.user.id.toLowerCase(),
+        ),
+    );
+  }
+
+  // get reputation for each address
+  const contributors = await Promise.all(
+    addressesWithReputation.map(async (address) => {
+      const { data } = await graphqlRequest(
+        getUser,
+        {
+          id: utils.getAddress(address),
+        },
+        GRAPHQL_URI,
+        API_KEY,
+      );
+      const user = data?.getUserByAddress?.items[0] || {};
+
+      try {
+        const userReputationForAllDomains =
+          await colonyClient.getReputationAcrossDomains(address, rootHash);
+
+        // Filter based on domainId if provided
+        const filteredReputationForAllDomains =
+          userReputationForAllDomains.filter(
+            (userReputation) =>
+              userReputation.reputationAmount &&
+              !userReputation.reputationAmount.isZero() &&
+              userReputation.domainId === domainId,
+          );
+
+        // Extract only the relevant data and
+        // transform the user reputation amount on each domain to percentage
+        const formattedUserReputations = filteredReputationForAllDomains.map(
+          (userReputation) => {
+            if (!userReputation?.reputationAmount) {
+              return {};
+            }
+
+            const reputationPercentage = calculatePercentageReputation(
+              userReputation.reputationAmount?.toString(),
+              totalColonyReputation.reputationAmount?.toString(),
+            );
+
+            return {
+              reputationPercentage,
+              reputationAmount: userReputation.reputationAmount?.toString(),
+            };
+          },
+        );
+
+        return {
+          user: user || {},
+          reputationPercentage:
+            formattedUserReputations[0]?.reputationPercentage,
+          reputationAmount: formattedUserReputations[0]?.reputationAmount,
+        };
```

Comment:
What happens if a address has reputation, but not a user account registered ?

This code will return it's reputation, but no way of identifying (hence displaying) said address, since the user object will be empty.

This is a thing that will come up normally, not even in edge cases, so this needs to be refactor to account for that.

Reply:
@rdig Should we maintain that difference? i.e. keep the address separate? or fill-in the address in the user object?

{
          user: user || {},
          reputationAddress, // store here
          reputationPercentage,
          reputationAmount:,
}

A separate address would require more attention to handle (check user first, then check for an address).

Is there a simple way to simulate this situation for testing so that i can be sure the its covered?

Reply:
I've opted to fill in the address/id in User.